lib/spack/spack/build_environment.py

```diff
@@ -162,75 +162,49 @@ def clean_environment():
     # unlike the other functions so it doesn't overwrite what the modules load.
     env = EnvironmentModifications()
 
-    # Remove these vars from the environment during build because they
-    # can affect how some packages find libraries.  We want to make
-    # sure that builds never pull in unintended external dependencies.
-    env.unset('LD_LIBRARY_PATH')
-    env.unset('LD_RUN_PATH')
-    env.unset('DYLD_LIBRARY_PATH')
-    env.unset('DYLD_FALLBACK_LIBRARY_PATH')
-
-    # These vars affect how the compiler finds libraries and include dirs.
-    env.unset('LIBRARY_PATH')
-    env.unset('CPATH')
-    env.unset('C_INCLUDE_PATH')
-    env.unset('CPLUS_INCLUDE_PATH')
-    env.unset('OBJC_INCLUDE_PATH')
-
-    env.unset('CMAKE_PREFIX_PATH')
-    env.unset('PYTHONPATH')
-
-    # Affects GNU make, can e.g. indirectly inhibit enabling parallel build
-    env.unset('MAKEFLAGS')
-
-    # Avoid that libraries of build dependencies get hijacked.
-    env.unset('LD_PRELOAD')
-    env.unset('DYLD_INSERT_LIBRARIES')
-
-    # Avoid <packagename>_ROOT user variables overriding spack dependencies
-    # https://cmake.org/cmake/help/latest/variable/PackageName_ROOT.html
-    # Spack needs SPACK_ROOT though, so we need to exclude that
-    for varname in os.environ.keys():
-        if varname.endswith('_ROOT') and varname != 'SPACK_ROOT':
-            env.unset(varname)
+    env_name_whitelist = list(map(re.compile, (
+        r'SPACK',
+        r'^SPACK_.*',  # keep SPACK_ROOT and company
+        r'EDITOR',
+        r'PATH',
+        r'PATHEXT',
+    )))
```

Comment:
Not clear to me that we should keep any of these except `SPACK_*`.  Is there some reason not to start PATH out empty or with some very small explicit set of paths?

Reply:
At the time, we relied so heavily on leaking that it seemed necessary, now I'm not so sure. It would be worth testing for sure.

Reply:
I think we should wall the whole env off and be purely constructive. Things can still leak in through externals, but that's mostly unavoidable.

Reply:
I agree, but we are missing _tons_ of dependencies on basic system tools like sed, awk, etc.  It might be a big lift if `/usr/bin` and `/bin` are gone.  Maybe we can include those for now and start tracking how many things like that are left?

Reply:
I think it's worth running a pipeline to see what happens... I agree we don't test everything but I wonder if we can handle a lot of these at the build tool level (e.g. for `autotools` packages).  I think we have to get them updated at some point.